version: "3.7"

x-hdfs-common:
  &hdfs-common
  image: gchq/hdfs:latest
  networks:
    default_net:
  environment:
    - CLUSTER_NAME=hadoop
  volumes:
    - ./config:/opt/hadoop/etc/hadoop:ro
  restart: on-failure:3


x-datanode-common:
  &datanode-common
  <<: *hdfs-common
  depends_on:
      - namenode
  command: datanode

services:
  namenode:
    <<: *hdfs-common
    container_name: namenode
    hostname: namenode
    ports:
      - "9870:9870"
      - "9000:9000"
    volumes:
      - ./config:/opt/hadoop/etc/hadoop:ro
      - ./data/namenode:/data1
    command: namenode

  datanode1:
    <<: *datanode-common
    container_name: datanode1
    hostname: datanode1
    volumes:
      - ./data/datanode1:/data1
      - ./config:/opt/hadoop/etc/hadoop:ro

  datanode2:
    <<: *datanode-common
    container_name: datanode2
    hostname: datanode2
    volumes:
      - ./data/datanode2:/data1
      - ./config:/opt/hadoop/etc/hadoop:ro
      
  datanode3:
    <<: *datanode-common
    container_name: datanode3
    hostname: datanode3
    volumes:
      - ./data/datanode3:/data1
      - ./config:/opt/hadoop/etc/hadoop:ro

  # datanode4:
  #   <<: *datanode-common
  #   container_name: datanode4
  #   hostname: datanode4
  #   volumes:
  #     - ./data/datanode4:/data1
  #     - ./config:/opt/hadoop/etc/hadoop:ro

  # datanode5:
  #   <<: *datanode-common
  #   container_name: datanode5
  #   hostname: datanode5
  #   volumes:
  #     - ./data/datanode5:/data1
  #     - ./config:/opt/hadoop/etc/hadoop:ro

volumes:
  namenode_data:
  datanode1_data:
  datanode2_data:
  datanode3_data:
  datanode4_data:
  datanode5_data:
    
networks:
  default_net:
    external: true
